### DISCUSSION TITLE
Does O3DE support binary installation plug-ins?

### DISCUSSION THREAD

**Discussion by southeast2017** [o3de/o3de#16427] [Q&A]
We have developed some plugins by O3DE, but we have to compile the plugins with the engine each time, which takes too long. In addition, the plugins need to be released with the engine. Does the engine support the binary mode of releasing plugins? In this way, the O3DE SDK can integrate the functions of the plugin by installing the plugin's binary file. 
Anyone know that？Thanks!

**Comment by lemonade-dm** [comment]
Yes, actually. 
We made an [Prebuilt Gem](https://github.com/o3de/o3de/tree/development/Templates/PrebuiltGem/Template) template that can contains placeholders that can be filed out with the binary shared and static libraries, as well as any executable.

What goes into a prebuilt Gem target is that it is a [CMake IMPORTED](https://cmake.org/cmake/help/latest/command/add_library.html#imported-libraries) target, where the [IMPORTED_LOCATION](https://cmake.org/cmake/help/latest/prop_tgt/IMPORTED_LOCATION.html#prop_tgt:IMPORTED_LOCATION) property is set to the library artifact and the [target_include_directories](https://cmake.org/cmake/help/latest/command/target_include_directories.html#command:target_include_directories) function is used to configure the include directories.

You can use the following command to instantiate the PrebuiltGem tempalte
```bash
<your-engine-root>$ scripts\o3de.bat create-gem --tn PrebuiltGem -dp <destination-path-for-your-gem/plugin>
```

Once that is done, the binary artifacts for the Gem(which is O3DE name for a Plugin Package) can be placed into the Gem folder and the `IMPORTED_LOCATION` properties with within the instantiated [per platform and build permutation](https://github.com/o3de/o3de/blob/development/Templates/PrebuiltGem/Template/Code/Platform/Windows/Default/%24%7BNameLower%7D.cmake#L11-L12) can be filled out with the library.

Since a Gem by definitions is just a folder containing a 1. `gem.json` file with a "gem_name" field in it and 2. a `CMakeLists.txt`, it means that the once Gem's `CMakeLists.txt` is called, the Gem author is in control of what occurs next.
Be that building code, referencing already built code via IMPORTED targets, or maybe performing a random operation such as using CMake `execute_process` command to do any action the Gem author desires.

The [README.md](https://github.com/o3de/o3de/blob/development/Templates/PrebuiltGem/Template/README.md) instructions for the PrebuiltGem templates explains how to use it.
If the instructions are insufficient, please let us know.

**Reply by southeast2017** [reply]
Hello, I don't mean plug-ins importing static libraries or dynamic libraries, I mean, how do we compile plug-ins into binary files, and then we can install the plug-ins into the compiled SDK in some way. Like VSCode, binary plug-ins are dynamically installed and uninstalled, rather than source-sourced.

**Reply by lemonade-dm** [reply]
If you register the a Gem with the engine.json using the o3de CLI python script `register` command, then when you build an SDK layout, it will compile the Gem into it's binary output as part of the SDK

Using the `--esep/--external-subdirectory-engine-path` option to the `register` command, the Gem can be forced registerd with the engine instead of the global manifest file in `$HOME/.o3de/o3de_manifest.json`
```bash
<your-engine-root>$ scripts\o3de.bat register --gem-path <YourGem> --external-subdirectory-engine-path
```

The documentation at https://www.docs.o3de.org/docs/user-guide/project-config/cli-reference/#usage-6 provides more in-depth explanation of the o3de CLI options

Once the Gem is registered with the engine, an engine-centric workflow can be used to build an SDK layout.

That involves passing the in the <your-engine-root> directory as the CMake SOURCE_DIR via the `-S` option.
Or for a simpler approach using a CMake Preset. 
The instructions for building an SDK layout is located at https://www.docs.o3de.org/docs/user-guide/build/distributable-engine/#build-the-install-target.

The general steps is to use CMake to generate a build solution that does NOT have an project associated with it.
This is so that the SDK is can be used with any O3DE after being built
```bash
# Generate a build solution. This generated in the folder of <your-engine-root>/windows
<your-engine-root>$ cmake --preset windows-default -DO3DE_INSTALL_ENGINE_NAME=YourCustomEngineName

# Use the `install`  target to build an SDK layout with only the profile configuration artifacts
cmake --build --preset windows-install --config profile
```

By default the SDK layout will be populated in `<your-engine-root>/install` folder

Optionally you can add additional build configurations to the SDK layout by building the `install` target with the `debug` or `release` configuration
```bash
# Add release artifacts to an existing SDK layout
cmake --build --preset windows-install --config release
```

Once that is done that install folder should contain the engine headers /binaries plus any Gem headers and binaries you registered with the engine before building the SDK layout
